[Terence Ho Wei Yang] iP

Vielheim opened the pull request.

Commit:
Add Gradle support

Vielheim committed:
Add Level-1 Requirements (Greet,Echo,Exit)

Vielheim committed:
Add Requirements for Level 2

Vielheim committed:
Add Requirements for Level 3

Vielheim committed:
Add Requirements for Level 4

Vielheim committed:
Add Requirements for Level 3

Vielheim committed:
Merge commit 'fb6c9205372c8a4ff484b4f51704d1ec15175bbd'

* commit 'fb6c9205372c8a4ff484b4f51704d1ec15175bbd':
  Add Requirements for Level 3

Vielheim committed:
Add Requirements for Automated Text UI Testing

Vielheim committed:
Add test case for "TODO" task (Empty Description)

Vielheim committed:
Add Requirements for Level 5

Vielheim committed:
Change DukeTask[] array into ArrayList<DukeTask> for invokeDuke()

Vielheim committed:
Add Requirements for Level 6

Vielheim committed:
Add Requirements for A-Enums (Where applicable)

Vielheim committed:
Add Requirements for Level 7

Vielheim committed:
Add Requirements for Level 8

Vielheim committed:
Bug Fixes for Level 8

Vielheim committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Bug Fixes for Level 8
  Add Requirements for Level 8

# Conflicts:
#	src/main/java/Duke.java

Vielheim committed:
Add Requirements for A-MoreOOP

Vielheim committed:
Add Requirements for A-Packages

Vielheim committed:
Add Requirements for A-JUnit

Vielheim committed:
Add Requirements for A-Jar

Vielheim committed:
Add Requirements for A-Javadoc

Vielheim committed:
Add Requirements for A-CodingStandard

Vielheim committed:
Add Requirements for Level 9

Vielheim committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add Requirements for A-Javadoc

Vielheim committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Add Requirements for A-CodingStandard

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/command/HelpCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/logic/InputValidator.java
#	src/main/java/duke/logic/UIManager.java

Vielheim committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add Requirements for Level 9

# Conflicts:
#	src/main/java/duke/logic/InputValidator.java
#	src/main/java/duke/logic/UIManager.java
#	src/main/java/duke/logic/UserInputParser.java

Vielheim committed:
Add more DukeException sub-classes

Expanded the initial DukeException class with more specific DukeException sub-classes to target more specific exceptions.
Simplified the design of the DukeException class to make it easier to manage.

Vielheim committed:
Merge branch 'branch-improve-exception'

* branch-improve-exception:
  Add more DukeException sub-classes

Vielheim committed:
Merge branch 'add-gradle-support' of https://github.com/Vielheim/ip into branch-A-Gradle

Vielheim committed:
Add Requirements for A-Gradle

Vielheim committed:
Merge branch 'branch-A-Gradle'

* branch-A-Gradle:
  Add Requirements for A-Gradle
  Add Gradle support

Vielheim committed:
Add Requirements for A-CheckStyle

Vielheim committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Add Requirements for A-CheckStyle

Jargonx commented on src/main/java/duke/CommonMethod.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Defines common methods used by Duke Program.
+ */
+public class CommonMethod {
```
Its a good idea to have a class of Common methods that are used throughout your code! I think I may implement this in my future projects!

Jargonx commented on src/main/java/duke/CommonString.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Denotes Common Strings used in <code>Duke</code>
+ */
+public enum CommonString {
```
Refer to comment on CommonMethod, is also applicable here!

Jargonx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,86 @@
+public class Duke {
+    private final UiManager uiManager;
+    private final StorageManager storageManager;
+    private TaskList taskList;
+
+    public Duke() {
+        this.uiManager = new UiManager();
+        this.storageManager = new StorageManager(CommonString.DUKE_FILE_PATH.toString());
+        try {
+            this.taskList = new TaskList(storageManager.loadData());
```
Perhaps you can think about shifting your taskList object to the StorageManager Class? Having your taskList in your main class could possibly complicate your manipulations of the items inside!

Jargonx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,86 @@
+                Command command = UserInputParser.parse(userInput);
+                command.execute(taskList, uiManager, storageManager);
+                isExit = command.getExitStatus();
+            } catch (UnknownInstructionException | InvalidInstructionFormatException
+                    | MissingFieldException | TaskDoneException
+                    | InvalidInstructionLengthException | InvalidTaskIndexException e) {
+                System.out.println(e);
+            } catch (IOException e) {
+                System.out.println("IO Error: " + e.getMessage());
+                e.printStackTrace();
+            }
```
Perhaps you can catch the IOException error within the originating class and throw a DukeException instead, that way the main Duke Class never has to deal with Exceptions that are expose the Java Exceptions! Although it is always good to be explicit in what errors you are catching, since most of the originating errors are Extended from DukeException, you could always just use 
```suggestion
            } catch (DukeException e) {
                System.out.println(e);
            }
```

Vielheim replied:
Thanks man! I'll take note for the IOExceptions!

Jargonx commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,49 @@
+ * It also contains an abstract method <code>execute</code> for implementing classes.
+ * Commands must have a method to execute their respective commands.
+ */
+public abstract class Command {
+
+    private final boolean isExit;
+
+    public Command(boolean isExit) {
+        this.isExit = isExit;
+    }
+
+    /**
+     * Returns a <code>Boolean</code> denoting if the given program should exit.
+     *
+     * @return boolean isExit.
+     */
+    public boolean getExitStatus() {
+        return isExit;
+    }
```
Perhaps there can be a better way to Store your isExit variable instead of having to pass it in to every Command Object that is initialised, since not every Command Object needs to know of the existence of the isExit parameter, as can be seen from your AddCommand class constructor, where you initialise isExit with false.

Vielheim replied:
Hmm that's true! That'll simplify it a lot

Jargonx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,86 @@
+            try {
+                String userInput = uiManager.readCommand();
+                Command command = UserInputParser.parse(userInput);
+                command.execute(taskList, uiManager, storageManager);
```
As with the taskList being passed around the entire program, the uiManager and storageManager Object are being passed around the entire program, which may not be necessary. Consider using a Non-Persistent Storage Object at initialisation of program so that you can easily reference to the non persistent data throughout the program without using up parameter space to pass in objects that may or may not be used by the methods/classes that they are being passed into

Jargonx commented on src/main/java/duke/logic/InputValidator.java:
```diff
@@ -0,0 +1,199 @@
+        }
+
+        // input validation for date
+        if (year <= 0 || month <= 0 || month > 12 || day <= 0) {
+            throw new InvalidInstructionFormatException();
+        }
+
+        switch (month) {
+        case 1:
+        case 3:
+        case 5:
+        case 7:
+        case 8:
+        case 10:
+        case 12:
+            if (day > 31) {
+                throw new InvalidInstructionFormatException();
+            }
+            break;
+        case 4:
+        case 6:
+        case 9:
+        case 11:
+            if (day > 30) {
+                throw new InvalidInstructionFormatException();
+            }
+            break;
+        default:
+            if ((CommonMethod.isLeapYear(year) && day > 29) || (!CommonMethod.isLeapYear(year) && day > 28)) {
+                throw new InvalidInstructionFormatException();
+            }
+        }
```
Consider using in built Date and Time objects that are provided by the Java JDK! They provide extensive functionality including parsing and Formatting abilities!

Jargonx commented on src/main/java/duke/logic/StorageManager.java:
```diff
@@ -0,0 +1,128 @@
+ * Represents the Storage of Duke.
+ * Manages the storing/loading of data from data file.
+ */
+public class StorageManager {
```
Perhaps you can consider splitting this class into multiple classes, such as a Save and Load file which deals with persistent storage and consider having another class for Non-Persistent storage! Making your classes more modular will help you maintain your code in the future!

Jargonx commented on src/main/java/duke/logic/StorageManager.java:
```diff
@@ -0,0 +1,128 @@
+     *
+     * @throws IOException If saving of the data fails
+     */
+    public void saveData(ArrayList<DukeTask> dataList) throws IOException {
```
Refer to comment on IOException in the main Duke Class!

Jargonx commented on src/main/java/duke/logic/TaskList.java:
```diff
@@ -0,0 +1,64 @@
+ * It contains an <code>ArrayList</code> to track the user's tasks
+ * and methods to operate on it.
+ */
+public class TaskList {
```
Perhaps instead of making this ArrayList Object into a class and only exposing the methods that are required, you can consider using a Storage class from which you can accomplish the same goal but have the flexibility of having more methods related to storage rather than restrict yourself to methods only relating to one specific data structure!

Jargonx commented on src/main/java/duke/logic/UiManager.java:
```diff
@@ -0,0 +1,199 @@
+ * It contains <code>Enums</code> for pre-defined texts to be printed for the program
+ * and methods to print feedback to the users for different commands.
+ */
+public class UiManager {
```
I like this class because its function has been very well defined and followed through with! Good job! 👍

Jargonx commented on src/main/java/duke/logic/UserInputParser.java:
```diff
@@ -0,0 +1,212 @@
+            InvalidInstructionFormatException, UnknownInstructionException {
+
+        // Instruction Setup - split by whitespace to check
+        String instruction = userInput.trim();
```
Perhaps you can consider another case - what happens if a user types in a command with the wrong capitalisation? Does your program still accept that as the correct command?

Jargonx commented on src/main/java/duke/task/DukeTask.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return String status icon.
+     */
+    public String getStatusIcon() {
+        return (isDone ? "\u2713" : "\u2718"); //return tick or X symbols
+    }
```
Its good that your UI has such a User friendly display! Maybe you can instead define these as private final constant variables above as class level members since these look like Magic Numbers that may confuse future reviewers!

Jargonx commented on src/main/java/duke/task/DeadlineTask.java:
```diff
@@ -0,0 +1,37 @@
+     *
+     * @return String DateTime.
+     */
+    public String getDatetime() {
+        DateTimeFormatter df = DateTimeFormatter.ofPattern(CommonString.DUKE_DATETIME_FORMAT.toString());
+        return datetime.format(df);
+    }
```
This seems like a common method for DeadlineTask and EventTask, is there a better place that you could place this method?

Jargonx commented on src/test/java/logic/TaskListTest.java:
```diff
@@ -0,0 +1,56 @@
+import duke.task.DukeTask;
+import stub.DukeTaskStub;
+
+public class TaskListTest {
```
You seem to have done extensive testing! Good job! 👍

Jargonx reviewed (commented):
Overall, your code seems to be very focused on the user experience which is a good way to go about doing things, especially since you have many tests, suggesting that you would really not want your code to run into any issues when your user is using your Duke Bot.

However, from a developer perspective perhaps you can take some time to look through the architecture of your code to try and understand the dependencies that you are creating by around certain objects or variables and reconsider the architecture based on the minimal functionality of each class and considering what classes in your ecosystem that they need to know about and which they dont.

Vielheim committed:
Add Requirements for Level-10 (In-Progress)

Finished first 3 levels of Level-10. Committing tfor revision control.

Vielheim committed:
Add last level changes of GUI tutorial

Vielheim committed:
Format DialogBox layout

Vielheim committed:
Implement Duke Response logic in GUI

Vielheim committed:
Add Gudetama Theme

Vielheim committed:
Modify text output to Gudetama theme

Vielheim committed:
Add window resizing functionality to GUI

bowei-yu reviewed (commented):
Overall, your PR is very easy to read, and I really like your coding style. Your code is extremely clean, one could hardly catch any coding standard violations. All of your Javadocs have detailed descriptions. I think I can learn many good practices from your work!

Vielheim committed:
Improve GUI

Vielheim committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Improve GUI
  Add window resizing functionality to GUI
  Modify text output to Gudetama theme
  Add Gudetama Theme
  Implement Duke Response logic in GUI
  Format DialogBox layout
  Add last level changes of GUI tutorial
  Add Requirements for Level-10 (In-Progress)

Vielheim commented:
Thanks @Jargonx and @bowei-yu for your reviews! Will take note and work on them :))

Vielheim committed:
Refactor Command method

Refactored the method getExitStatus() into an abstract method and removed the boolean isExit.
Reason being it doesn't serve much purpose other than being used in the method.

Vielheim committed:
Refactor DukeTask

Extract out the common time formatting functionalities for DeadlineTask and EventTask into a new abstract class DukeTaskWithTime

Vielheim committed:
Apply CheckStyle on GUI package

Vielheim committed:
Refactor Input Validator

Extracted the functionalities that validates Date and Time in validateDateAndTime() into two separate methods:
validateDate() and validateTime() to better manage their complexities.

Vielheim committed:
Extract UiManager functionalities into Interfaces

Extracted the functionalities pertaining to Intro/Outro/General User interaction to UserInteractionUi interface
Extracted the functionalities pertaining to Command Response interaction to CommandInteractionUi interface
Implemented interfaces in UiManager
Refactored classes that reference to UiManager with the interfaces
Done so to prevent miscalling of certain methods in said classes

Vielheim committed:
Apply CheckStyle to packages

Vielheim committed:
Refactor DukeExceptions

Added two new Exceptions:
DukeIoException and DukeFileNotFoundException
Encapsulates Java Exceptions so they are hidden from the client/implementation.
Recommended by @Jargonx

Vielheim committed:
Update build.gradle to allow runShadow

Vielheim committed:
Add Requirements for A-Assertions

Vielheim committed:
Add Requirements for A-CodeQuality

Vielheim committed:
Improve Code Quality

Vielheim committed:
Minor Bug Fix

FindCommand has the wrong getExitStatus() method

Vielheim committed:
Add Requirements for A-Streams

Where applicable and does not compromise on code readability

Vielheim committed:
Merge pull request #1 from Vielheim/branch-A-Assertions

Add Requirements for A-Assertions

Vielheim committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add Requirements for A-Assertions

# Conflicts:
#	src/main/java/duke/logic/InputValidator.java
#	src/main/java/duke/logic/StorageManager.java

Vielheim committed:
Merge branch 'master' into branch-A-Streams

* master:
  Add Requirements for A-Assertions

Vielheim committed:
Merge pull request #3 from Vielheim/branch-A-CodeQuality

Add Requirements for A-CodeQuality

Vielheim committed:
Merge branch 'master' into branch-A-Streams

* master:
  Improve Code Quality
  Add Requirements for A-CodeQuality

Vielheim committed:
Merge pull request #2 from Vielheim/branch-A-Streams

Add Requirements for A-Streams

Vielheim committed:
Add Requirements for A-Lambdas

Vielheim committed:
Merge pull request #4 from Vielheim/branch-A-Lambdas

Add Requirements for A-Lambdas

Vielheim committed:
Add Requirements for A-CI

Vielheim committed:
Improve JUnit Testing

Rewriting code for JUnit Testing to test better

Vielheim committed:
Add more JUnit Tests

Vielheim committed:
Add more JUnit Tests

Vielheim committed:
Finalise addition of JUnit Tests

Vielheim committed:
Add Requirements for BCD-Extensions

1. C-Help (Already implemented in early stages)
2. C-BetterSearch (Able to find partial keywords, NOT case-sensitive)

BetterSearch uses a lowercase keyword contains on a lowercase String to be able to search for partial keywords

Vielheim committed:
Add Requirements for BCD-Extensions

C-Sort:
Allows the user to sort their TaskList and display sorted Tasks

Vielheim committed:
Merge branch 'branch-BCD-Extensions'

* branch-BCD-Extensions:
  Add Requirements for BCD-Extensions
  Add Requirements for BCD-Extensions

Vielheim committed:
Bug fixes (test)

Attempt to fix bug where the Github CI won't pass on build (Windows-Latest)
Change JUnit Test where the test failed (mainCommandTest)

Vielheim committed:
Refactor code according to A-CodeQuality (SLAP)

Vielheim committed:
Add Requirements for A-BetterGui

Vielheim committed:
Update README.md

Vielheim committed:
Update README.md

Vielheim committed:
Update README.md

Vielheim committed:
Fix CheckStyle Errors

Vielheim committed:
Update README.md

Vielheim committed:
Add Requirements for A-UserGuide

Vielheim committed:
Update README.md

Vielheim committed:
Add Title to GUI window

Vielheim committed:
Update README.md

Vielheim committed:
Update README.md
